### ISSUE TITLE
Passing config.settings to plugin processors

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#5136]
I was hoping we could add `config.settings` to processors (e.g., pass them in on `preprocess()` and `postprocess()`). Right now, those settings are only accessible in rules.

One way I could see it being done is adding a third argument to those function contracts. To avoid further argument creep, though, it might be worth creating a sort of "pluginContext" or "processorContext" object that conceptually parallels RuleContext, and then we could add more properties there.

In any case, adding a third argument to those functions should be backward-compatible, so the challenge in that case becomes freezing or defensively-cloning `config.settings` in CLIEngine to avoid malicious modifications. But there might be a better approach.

**Comment by eslintbot** [comment]
@platinumazure Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
Third argument sounds reasonable to me. Although, I would make it an object with settings as a key. If we need it, we can make it into an pluginContext later.

**Comment by platinumazure** [comment]
@ilyavolodin Fine by me. Whether the object will later have methods or not, and whether the object was constructed via a constructor or via `{}`/`Object.create(null)` doesn't make much difference from inside the plugin.

**Comment by platinumazure** [comment]
Actually, folks, I've run into a use case where I also want `settings` (or some other configuration object) in order to determine what file extensions I actually want to provide my processor for. Is that possible now? And if not, should we pass in settings to plugins (...somehow) and then a processor can just get at those settings?

**Comment by ilyavolodin** [comment]
@platinumazure Not sure I'm following you. Can you explain in more details what you want to do?

**Comment by platinumazure** [comment]
@ilyavolodin Yes, absolutely, sorry about that.

Currently, a plugin exports `rules` and `processors`:

```js
module.exports = {
    rules: {
        // ...
    },
    processors: {
        ".txt": someProcessor,
        ".html": someOtherProcessor
    }
};
```

As you can see, the plugin is expected to export an object, so it has no way of accessing settings at that level. This implies that the extensions the plugin will process must also be static/constant.

What I'm hoping to do is allow users to specify what extensions they want covered in my plugin, and then export processors based off of that.

This is a separate issue from what I originally submitted, but I wanted to mention it here in case this new issue informs the direction of how we want to fix the original issue. For example, if we decide to allow plugins to export a function or constructor that could take in settings, it might make sense to just let the plugins pass the settings into the processor without changing that contract.

Let me know if I can make anything more clear, and/or if you want me to open a separate issue for the above scenario.

**Comment by ilyavolodin** [comment]
Yes, I think this is a separate issue. Basically you want to be able to create dynamic processors. I'm not sure that's a good idea, due to the fact you might have multiple processors fight for the same files. I think this might be outside of the scope for processors.

**Comment by platinumazure** [comment]
@ilyavolodin I'll open a separate issue. I don't agree with your rationale about it being "not [...] a good idea" because that's already a possibility as it is (ESLint picks the first plugin that accepts the file extension, and no, this isn't documented), but I'll take further discussion to that new issue. Thanks!

**Comment by platinumazure** [comment]
Okay, opened #5155 for the plugin extension issue. The proposal put forward by @nzakas on that issue means the third parameter with `settings` on it will work satisfactorily for this issue.

**Comment by platinumazure** [comment]
@nzakas Apologies for pinging you again, but do you think the third-parameter approach will work here? Or do we need to discuss other options before accepting the issue? Thanks!

**Comment by nzakas** [comment]
If we do #5155, is this still necessary?

**Comment by platinumazure** [comment]
IMO yes. #5155 doesn't have anything to do with config.settings being available in plugins. I had assumed I would need config.settings for extension management, but you proposed having ESLint reclaim control of extensions, which makes sense.

My use case here has more to do with knowing if certain rules' errors should be ignored by my plugin, which I think is still a use case for config.settings. I don't need config.settings injected into the main plugin file (which would require integrating with plugins as function as well as objects), but I still think there's a use for allowing `preprocess` and `postprocess` to have access to config.settings just like rules do.

**Comment by nzakas** [comment]
> My use case here has more to do with knowing if certain rules' errors should be ignored by my plugin, which I think is still a use case for config.settings. I

Can you explain more about your use case?

**Comment by platinumazure** [comment]
@nzakas Actually, sorry, that's a bad example because glob configuration would solve the use case I presented. (What I was getting at was it would be nice to allow users to suppress certain rules' messages simply because they made no sense for microtemplates. But if rules and plugin processors can be configured via glob configuration, then users can simply disable rules directly.)

Here's a better example: In microtemplates, we need to match on certain delimiters in order to know what is text and what is code. Underscore and Lodash use ERB delimiters by default, and that works for probably 80+% of the plugin's eventual users, but if a user wants to use different delimiters, the `preprocess` function (in this case) needs to have the desired delimiters passed in as settings. I don't see how glob configuration helps with that, except in that the relevant settings can be kept out of the configuration for other files.

Does that make sense?

**Comment by nzakas** [comment]
Hmm, I think we need to think more about this. Right now, your suggestion is to piggyback on top of settings as a way to pass data into plugins. We should think through if that's the right way to pass information to plugins or if there's another approach that is better suited to this use case.

**Comment by kewisch** [comment]
I have another use case for this feature. We are preprocessing some files from xml to javascript, where some of the xml nodes contain javascript. The result looks something like this:

```
<bindings id="foo" xmlns="http://www.mozilla.org/xbl">
<binding id="bar">
    <implementation>
      ...
      <property name="occurrence" readonly="true">
        <getter><![CDATA[
            return this.mOccurrence;
        ]]></getter>
      </property>
    </implementation>
  </binding>
</bindings>
```
```javascript
this.bindings = {
  "bar": {
    implementation: {
      get occurrence() {
          return this.mOccurrence;
      }
    }
  }
};
```
Now the problem is when I add the `indent` rule to be 4-space, then it fails for most lines, pointing to somewhere in the original xml. I would like to use the config to detect if an indent rule is specified and if so use four spaces instead of two.

**Comment by nzakas** [comment]
@kewisch that's a case we explicitly do not want to support. Processors should not know what rules are enabled or what the rule options are.

**Comment by kewisch** [comment]
I understand, but how would you propose to solve the use case above? I am not interested in changing the indent of the line for `return this.mOccurrence;` within the preprocessor, but I need to have a value for the supporting lines around that, and with the current config this is not something the developer can influence.

So if we do what we have now to assume 2 space indent, but part of the tree preferring 4 space indent, then those using 4 space will just be out of luck with the indent rule and will have to live with warnings or disable the rule.

**Comment by nzakas** [comment]
@kewisch we never claimed to support that use case. The role of processors is transform the code into a format that ESLint can understand and then pass back the results so you can munge them however you'd like. Processors will never be allowed to know rule configurations as they are not designed to care about them. Perhaps you want to outdent your code in the processor before passing it to ESLint so you get the correct results?

In any event, this is tangential to the point of this issue. If you have more questions about processors, please stop by the chatroom

**Comment by nzakas** [comment]
Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
